odoo/tools/migration.py

```diff
@@ -0,0 +1,2383 @@
+               FROM res_groups_users_rel
+           GROUP BY uid
+             HAVING array_agg(gid) @> %s
+             EXCEPT
+             SELECT uid, gid
+               FROM res_groups_users_rel
+              WHERE gid = %s
+    """, [to_group, from_groups, to_group])
+
+# Models & Fields utilities
+
+
+def create_m2m(cr, m2m, fk1, fk2, col1=None, col2=None):
+    """Create a many2many relation table.
+
+        :param str m2m: relation table name
+        :param str fk1: table referenced by first column
+        :param str fk2: table referenced by second column
+        :param str col1: first column name
+        :param str col2: second column name
+    """
+    if col1 is None:
+        col1 = '%s_id' % fk1
+    if col2 is None:
+        col2 = '%s_id' % fk2
+
+    cr.execute("""
+        CREATE TABLE {m2m}(
+            {col1} integer NOT NULL REFERENCES {fk1}(id) ON DELETE CASCADE,
+            {col2} integer NOT NULL REFERENCES {fk2}(id) ON DELETE CASCADE,
+            UNIQUE ({col1}, {col2})
+        );
+        CREATE INDEX ON {m2m}({col1});
+        CREATE INDEX ON {m2m}({col2});
+    """.format(**locals()))
+
+
+def ensure_m2o_func_field_data(cr, src_table, column, dst_table):
+    """Fix broken many2one relations.
+
+        If any `column` not present in `dst_table`, remove column from `src_table` in
+        order to force recomputation of the function field
+
+        WARN: only call this method on m2o function/related fields!!
+    """
+    if not column_exists(cr, src_table, column):
+        return
+    cr.execute("""SELECT count(1)
+                    FROM "{src_table}"
+                   WHERE "{column}" NOT IN (SELECT id FROM "{dst_table}")
+               """.format(src_table=src_table, column=column, dst_table=dst_table))
+    if cr.fetchone()[0]:
+        remove_column(cr, src_table, column, cascade=True)
+
+
+def uniq_tags(cr, model, uniq_column='name', order='id'):
+    """Deduplicate "tag" models entries.
+
+        Should only be referenced as many2many
+
+        By using `uniq_column=lower(name)` and `order=name`
+        you can prioritize tags in CamelCase/UPPERCASE.
+    """
+    table = table_of_model(cr, model)
+    upds = []
+    for ft, fc, _, da in get_fk(cr, table):
+        assert da == 'c'    # should be a ondelete=cascade fk
+        cols = get_columns(cr, ft, ignore=(fc,))[0]
+        assert len(cols) == 1   # it's a m2, should have only 2 columns
+
+        upds.append("""
+            INSERT INTO {rel}({c1}, {c2})
+                 SELECT r.{c1}, d.id
+                   FROM {rel} r
+                   JOIN dups d ON (r.{c2} = ANY(d.others))
+                 EXCEPT
+                 SELECT r.{c1}, r.{c2}
+                   FROM {rel} r
+                   JOIN dups d ON (r.{c2} = d.id)
+        """.format(rel=ft, c1=cols[0], c2=fc))
+
+    assert upds         # if not m2m found, there is something wrong...
+
+    updates = ','.join('_upd_%s AS (%s)' % x for x in enumerate(upds))
```

Comment:
What should the first `%s` originally take as a value? - I doubt the whole query... :wink:

Reply:
you might want to recheck what enumerate does then; your faith is not required :stuck_out_tongue: 
``` SQL
_upd_0 AS (INSERT INTO ...etc
```

Reply:
Where is this simely banging my hand against the front. Thank you ;-)